src/Avalonia.Base/AvaloniaObject.cs

```diff
@@ -553,19 +559,44 @@ protected virtual void OnPropertyChanged(AvaloniaPropertyChangedEventArgs e)
         protected bool SetAndRaise<T>(AvaloniaProperty<T> property, ref T field, T value)
         {
             VerifyAccess();
-            if (!object.Equals(field, value))
+            if (!directDelayedSetter.IsNotifying(property))
             {
-                var old = field;
-                field = value;
-                RaisePropertyChanged(property, old, value, BindingPriority.LocalValue);
-                return true;
+                var valueChanged = false;
+                if (!object.Equals(field, value))
+                {
+                    SetAndRaiseCore(property, ref field, value);
+
+                    valueChanged = true;
+                }
+
+                while (directDelayedSetter.HasPendingSet(property))
+                {
+                    SetAndRaiseCore(property, ref field, (T)directDelayedSetter.GetFirstPendingSet(property));
+                    valueChanged = true;
+                }
+                return valueChanged;
             }
-            else
+            else if(!object.Equals(field, value))
+            {
+                directDelayedSetter.AddPendingSet(property, value);
+            }
+            return false;
+        }
+
+        private void SetAndRaiseCore<T>(AvaloniaProperty<T> property, ref T field, T value)
+        {
+            var old = field;
+            field = value;
+
+            using (directDelayedSetter.MarkNotifying(property))
             {
-                return false;
+                RaisePropertyChanged(property, old, value, BindingPriority.LocalValue);
             }
         }
 
+        protected void SetAndRaise<T>(AvaloniaProperty<T> property, Action<T, Action<Action>> setterCallback, T value, Predicate<T> pendingSetCondition = null)
+            => directDelayedSetter.SetAndNotify(property, (val, notify) => setterCallback((T)val, notify), value, o => pendingSetCondition?.Invoke((T)o) ?? true);
```

Comment:
Nit: I can't see the end of these lines on my laptop screen or on GitHub. The [Contributing document](http://avaloniaui.net/guidelines/contributing.html) says:

> Try to keep lines of code around 100 characters in length or less, though this is not a hard limit. If you're a few characters over then don't worry too much.

It makes reviewing hard - I have a 13in monitor on my laptop. I think expression bodied members can be abused ;)

Reply:
I'll fix the super-long lines.